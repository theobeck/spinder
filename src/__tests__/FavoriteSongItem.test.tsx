/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import FavoritesSongItem from "../components/FavoriteSongItem";
import '@testing-library/jest-dom'

describe('FavoriteSongItem Component', () => {
    const mockOnUnlike = jest.fn();

    const defaultProps = {
        albumCover: 'test-album-cover.jpg',
        title: 'Test Song Title',
        artist: 'Test Artist',
        songId: '12345',
        onUnlike: mockOnUnlike,
    };

    beforeEach(() => {
        render(
            <Router>
                <FavoritesSongItem {...defaultProps} />
            </Router>
        );
    });

    test('should render correctly (snapshot test)', () => {
        const tree = render(
            <Router>
                <FavoritesSongItem {...defaultProps} />
            </Router>
        );
        expect(tree).toMatchSnapshot();
    });
    test('should render album cover image with correct src and alt attributes', () => {
        const imgElement = screen.getByRole('img', { name: /song 1/i });
        expect(imgElement).toBeInTheDocument();
        expect(imgElement).toHaveAttribute('src', defaultProps.albumCover);
        expect(imgElement).toHaveAttribute('alt', 'Song 1');
    });

    test('should render correct song title', () => {
        expect(screen.getByText(defaultProps.title)).toBeInTheDocument();
    });

    test('should render correct artist name', () => {
        expect(screen.getByText(defaultProps.artist)).toBeInTheDocument();
    });

    test('should render unlike button and call onUnlike prop when clicked', () => {
        const unlikeButton = screen.getByRole('button', { name: /ðŸ—™/i });
        expect(unlikeButton).toBeInTheDocument();

        fireEvent.click(unlikeButton);
        expect(mockOnUnlike).toHaveBeenCalled();
    });

    test('should render link with correct href', () => {
        const linkElement = screen.getByRole('link');
        expect(linkElement).toHaveAttribute('href', `/track/${defaultProps.songId}`);
    });
});
