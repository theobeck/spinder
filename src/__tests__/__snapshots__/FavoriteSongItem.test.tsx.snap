// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteSongItem Component should render correctly (snapshot test) 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="song-item"
      >
        <a
          href="/track/12345"
        >
          <img
            alt="Song 1"
            src="test-album-cover.jpg"
          />
        </a>
        <div
          class="song-details"
        >
          <p
            class="song-title"
          >
            Test Song Title
          </p>
          <p
            class="song-artist"
          >
            Test Artist
          </p>
        </div>
        <div
          class="song-options"
        >
          <button
            class="song-unlike-button"
          >
            ðŸ—™
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="song-item"
      >
        <a
          href="/track/12345"
        >
          <img
            alt="Song 1"
            src="test-album-cover.jpg"
          />
        </a>
        <div
          class="song-details"
        >
          <p
            class="song-title"
          >
            Test Song Title
          </p>
          <p
            class="song-artist"
          >
            Test Artist
          </p>
        </div>
        <div
          class="song-options"
        >
          <button
            class="song-unlike-button"
          >
            ðŸ—™
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="song-item"
    >
      <a
        href="/track/12345"
      >
        <img
          alt="Song 1"
          src="test-album-cover.jpg"
        />
      </a>
      <div
        class="song-details"
      >
        <p
          class="song-title"
        >
          Test Song Title
        </p>
        <p
          class="song-artist"
        >
          Test Artist
        </p>
      </div>
      <div
        class="song-options"
      >
        <button
          class="song-unlike-button"
        >
          ðŸ—™
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
